phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,459 @@
+                LOGGER.severe("Something went wrong while checking and waiting for Major Compaction for table " + table+e);
+            }
+        });
+    }
+
+    private static void resetMajorCompaction(Admin admin, Map.Entry<String, ArrayList<String>> value) {
+        value.getValue().parallelStream().forEach(table -> {
+            try {
+                tableMC.put(table, admin.getDescriptor(TableName.valueOf(table)).isCompactionEnabled());
+                if (!dryRun) {
+                    TableDescriptorBuilder tdb = TableDescriptorBuilder.newBuilder(admin.getDescriptor(TableName.valueOf(table)));
+                    tdb.setCompactionEnabled(false);
+                    admin.modifyTable(tdb.build());
+                }
+                LOGGER.info("Major compaction disabled for "+table);
+
+            } catch (IOException e) {
+                LOGGER.severe("Something went wrong while setting Major Compaction to false for "+table+" "+e);
+            }
+        });
+    }
+
+    private static void setMajorCompaction(Admin admin, Map.Entry<String, ArrayList<String>> value) {
+        value.getValue().parallelStream().forEach(table -> {
+            try {
+                if (!dryRun) {
+                    TableDescriptorBuilder tdb = TableDescriptorBuilder.newBuilder(admin.getDescriptor(TableName.valueOf(table)));
+                    tdb.setCompactionEnabled(tableMC.get(table));
+                    admin.modifyTable(tdb.build());
+                }
+                LOGGER.info("Major compaction set to original value {"+tableMC.get(table)+"} for "+table);
+            } catch (IOException e) {
+                LOGGER.severe("Something went wrong while setting Major Compaction to old value "+table+" "+e);
+            }
+        });
+    }
+
+    private static void rebuildIndexes(Map.Entry<String, ArrayList<String>> value, Configuration conf) {
+        String schema = value.getKey().split("\\.")[0];
+        String table = value.getKey().split("\\.")[1];
+
+        value.getValue().stream().forEach(index -> {
+            IndexTool indexingTool = new IndexTool();
+            conf.set(QueryServices.TRANSACTIONS_ENABLED, Boolean.TRUE.toString());
+            indexingTool.setConf(conf);
+            ArrayList<String> argArray = new ArrayList();
+            argArray.add("-s");
+            argArray.add(schema);
+            argArray.add("-dt");
+            argArray.add(table);
+            argArray.add("-it");
+            argArray.add(index.split("\\.")[1]);
+            // Need to run this job in foreground for the upgrade to be deterministic
+            argArray.add("-direct");
+            argArray.add("-runfg");
+            argArray.add("-op");
+            argArray.add("/tmp/" + UUID.randomUUID().toString());
+            try {
+                LOGGER.info("Rebuilding index "+ index);
+                if(!dryRun) {
+                    indexingTool.run(argArray.toArray(new String[argArray.size()]));
+                }
```

Comment:
Does indexingTool wait for the index rebuild to finish? If not, is there another way to wait?

Reply:
Yes, I am passing -runfg (run foreground) which makes sure that the program waits till the tool completes. https://github.com/apache/phoenix/blob/master/phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexTool.java#L690